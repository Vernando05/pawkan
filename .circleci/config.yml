commands:
  create_cert:
    description: |
      Create certificates
    steps:
      - run:
          name: "Create certificates"
          command: |
            echo "$SERVERCRT" | base64 -d >> server.crt
            echo "$SERVERKEY" | base64 -d >> server.key
  build:
    description: |
      Create a production build
    steps:
      - run:
          name: "Create a production build"
          command: |
            NODE_ENV=production npm run build
            tar -cvzf build_artifacts.tar.gz dist
  test:
    description: |
      Test using Jest and Cypress
    steps:
      - run:
          name: "Jest unit testing"
          command: |
            npm run test:unit -- --collectCoverage
      - run: 
          name: "Cypress e2e testing"
          command: npm run test-ci
      - codecov/upload:
          file: coverage/clover.xml
  deploy:
    description: |
      Deploy to Firebase hosting
    steps:    
      - run: ./node_modules/.bin/firebase deploy --token "$FIREBASE_TOKEN" --only hosting
  lint:
    description: |
      Lint
    steps:
      - run:
          name: "Lint files"
          command: npm run lint
  install:
    description: |
      Install dependencies
    steps:
      - run:
          name: "Install dependencies"
          command: npm ci
orbs:
  codecov: codecov/codecov@1.1.3
executors:
  default:
    description: |
      Node 12 runtime environment
    docker:
      - image: cypress/browsers:<<parameters.cypress>>
    parameters:
      cypress:
        default: node12.13.0-chrome78-ff70
        type: string
jobs:
  build:
    description: |
      Build the application in production mode. The built assets are stored in artifacts.
    executor: default
    steps:
      - attach_workspace:
          at: ~/
      - create_cert
      - build
      - persist_to_workspace:
          paths:
            - project
          root: ~/
      - store_artifacts:
          path: ./build_artifacts.tar
  test:
    description: |
      Run the tests and generate coverage reports.
    executor: default
    steps:
      - attach_workspace:
          at: ~/
      - create_cert
      - test
  deploy:
    description: |
      Deploy to production
    executor: default
    steps:
      - attach_workspace:
          at: ~/
      - deploy
  lint:
    description: |
      Run lint on the source code
    executor: default
    steps:
      - attach_workspace:
          at: ~/
      - lint
  install:
    description: |
      Install dependencies. Checkout is called before install. (Optional) To prevent checkout from being called, pass your steps to before_install.
    executor: default
    parameters:
      before_install:
        default:
          - checkout
        description: |
          Steps to be executed before restoring cache and running install. Default is to run 'checkout' command.
        type: steps
    steps:
      - steps: <<parameters.before_install>>
      - restore_cache:
          key: v1-node-modules-{{ checksum "package-lock.json" }}
      - install
      - persist_to_workspace:
          paths:
            - project
            - .cache/Cypress
          root: ~/
      - save_cache:
          key: v1-node-modules-{{ checksum "package-lock.json" }}
          paths:
            - ~/.npm
            - ~/.cache
version: 2.1
workflows:
  version: 2.1
  front-test-build-deploy:
    jobs:
      - install
      - test:
          requires:
            - install
      - build:
          filters:
            branches:
              only:
                - dev
          requires:
            - test
      - deploy:
          filters:
            branches:
              only:
                - dev
          requires:
            - build
