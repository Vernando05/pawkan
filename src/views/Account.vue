<template>
  <div>
    <v-row v-if="!customerAccessToken" justify="center">
      <v-col>
        <SignInProvider class="mb-2" />
      </v-col>    
    </v-row>
    <v-btn v-else @click="destroyAccessToken" min-width="100%" outlined>{{ $t('signOut') }}</v-btn>
  </div>
</template>

<script lang="ts">  
  import mixins from 'vue-typed-mixins'
  import SignInProvider from '@/components/SignInProvider.vue'
  import { mapGetters, mapMutations } from 'vuex'
  import signIn from '@/mixins/signIn'
  import { UserError } from "shopify-storefront-api-typings"
  import clientStorage from '@/plugins/vue-cookies'
  import { CustomerUserError, CustomerAccessTokenCreatePayload } from 'shopify-storefront-api-typings'
  import customer from '@/graphql/Customer.gql'
  import customerAccessTokenCreateWithMultipass from '@/graphql/CustomerAccessTokenCreateWithMultipass.gql'
  import customerAccessTokenDelete from '@/graphql/CustomerAccessTokenDelete.gql'

  export default mixins(signIn).extend({
    name: 'Account',
    components: {
      SignInProvider
    },
    metaInfo () {
      return {
        title: this.$t('account'),
        titleTemplate: '%s | Pawkan - Your best pet care'      
      }    
    },
    data: () => ({
      errorMessages: [] as CustomerUserError[]      
    }),
    computed: {
      ...mapGetters([
        'customerAccessToken',
        'currentLanguage',
        'prevNav'
      ])
    },
    apollo: {
      customer: {
        query: customer,
        variables () {
          return {
            customerAccessToken: this.customerAccessToken
          }
        }
      }
    },       
    methods: {
      ...mapMutations([
        'removeCustomerAccessToken'
      ]),
      async createUser (): Promise<void> {
        this.errorMessages = []
        try {
          const result = await this.$apollo.mutate({
            mutation: customerAccessTokenCreateWithMultipass,
            variables: {
              multipassToken: ''
            }
          })          
          const { customerAccessTokenCreate: { customerAccessToken, customerUserErrors } }: { customerAccessTokenCreate: CustomerAccessTokenCreatePayload } = result.data
          if (customerUserErrors.length) {
            this.errorMessages = customerUserErrors
          } else
          if (customerAccessToken) {            
            clientStorage.removeItem('customerAccessToken')
            clientStorage.removeItem('customerAccessTokenExpiresAt')
            clientStorage.setItem('customerAccessToken', customerAccessToken.accessToken )
            clientStorage.setItem('customerAccessTokenExpiresAt', customerAccessToken.accessToken )
            this.$router.push(this.prevNav)
          }   
        } catch(error) {
          console.log(error)
          alert(error)
        }
      },
      async destroyAccessToken (): Promise<void> {
        try {
          const result = await this.$apollo.mutate({
            mutation: customerAccessTokenDelete,
            variables: {
              customerAccessToken: this.customerAccessToken
            }
          })
          const { userErrors }: { userErrors: UserError[] } = result.data.customerAccessTokenDelete
          if (userErrors.length) {
            console.log(userErrors)
          } else {
            // this.removeCustomerAccessToken()
          }
        } catch (error) {
          console.log(error)
          alert(error)
        }
      }
    }
  })
</script>
