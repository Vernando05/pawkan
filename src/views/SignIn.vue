<template>
  <div>
    <div v-if="errorMessages.length">
      <v-alert :key="index" v-for="(el, index) in errorMessages" type="error">{{ el.message }}</v-alert>
    </div>
    <ValidationObserver ref="observer" v-slot="{ invalid }">
      <form @submit.prevent="submit" role="form">
        <ValidationProvider v-slot="{ errors }" :name="$t('email')" rules="required|email">
          <v-text-field
            v-model="email"
            :error-messages="errors"
            :label="$t('email')"
            required
          ></v-text-field>
        </ValidationProvider>
        <ValidationProvider v-slot="{ errors }" :name="$t('password')" rules="required|min:6|max:10">
          <v-text-field
            v-model="password"
            :counter="10"
            :error-messages="errors"
            :label="$t('password')"
            type="password"
            autocomplete
            required
          ></v-text-field>
        </ValidationProvider>
        <v-btn :disabled="invalid" class="mr-4 mt-4" type="submit">{{ $t('submit') }}</v-btn>
      </form>
    </ValidationObserver>
  </div>
</template>

<script lang="ts">
  import store from '@/store'
  import mixins from 'vue-typed-mixins'
  import { mapGetters, mapMutations } from 'vuex'
  import { i18n } from '@/plugins/i18n'
  import { required, email, min, max } from 'vee-validate/dist/rules'
  import { extend, ValidationObserver, ValidationProvider, setInteractionMode } from 'vee-validate'
  import startCase from 'lodash/startCase'  
  import signIn from '@/mixins/signIn'
  import clientStorageResetToken from '@/mixins/clientStorageResetToken'
  import { CustomerUserError } from '@/types/shopify-storefront'

  setInteractionMode('aggressive')
  
  extend('required', {
    ...required,
    message: '{_field_} ' + i18n.t('cannotBeEmpty').toLowerCase()
  })

  extend('min', {
    ...min,
    message: '{_field_} ' + i18n.t('mustAtLeastHave', ['{length}']).toLowerCase()
  })

  extend('max', {
    ...max,
    message: '{_field_} ' + i18n.t('mayNotBeGreater', ['{length}']).toLowerCase()
  })

  extend('email', {
    ...email,
    message: i18n.t('emailMustBeValid'),
  })

  export default mixins(signIn, clientStorageResetToken).extend({
    name: 'SignIn',
    components: {
      ValidationProvider,
      ValidationObserver
    },
    metaInfo () {
      return {
        title: this.$t('signIn'),
        titleTemplate: '%s | Pawkan - Your best pet care'      
      }    
    },
    data: () => ({
      email: '' as string,
      password: '' as string,
      errorMessages: [] as CustomerUserError[]
    }),
    computed: {
      ...mapGetters([
        'prevNav'
      ])
    },   
    methods: { 
      ...mapMutations([
        'setCustomerAccessToken'
      ]),     
      textCapitalize (text: string): string {
        return startCase(text)
      },
      async submit (): Promise<void> {
        const isValid = await (this.$refs.observer as InstanceType<typeof ValidationProvider>).validate()
        if (isValid) {
          const result = await this.signIn_createToken({ email: this.email, password: this.password })
          if (result.customerUserErrors.length) {
            this.errorMessages = result.customerUserErrors
          } else
          if (result.customerAccessToken) {
            this.setCustomerAccessToken(result.customerAccessToken.accessToken)
            this.clientStorageResetToken_reset(result.customerAccessToken.accessToken)            
            this.email = ''
            this.password = '';
            (this.$refs.observer as InstanceType<typeof ValidationProvider>).reset()
            this.$router.push(this.prevNav)
          }
        }        
      }
    },
    beforeRouteEnter (...[, , next]) {
      if (store.getters.customerAccessToken) {
        next({ name: 'Home', params: { lang: 'en' } })
      } else {
        next()
      }
    }
  })
</script>