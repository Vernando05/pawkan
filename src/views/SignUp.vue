<template>
  <div>
    <FormProfile ref="formProfile" :submit="submit" :error="errorMessages" />
  </div>
</template>

<script lang="ts">
  import store from '@/store'
  import mixins from 'vue-typed-mixins'  
  import { mapGetters, mapMutations } from 'vuex'
  import FormProfile from '@/components/FormProfile.vue'
  import signUp from '@/mixins/signUp'
  import signIn from '@/mixins/signIn'
  import clientStorageResetToken from '@/mixins/clientStorageResetToken'
  import { CustomerCreateInput, CustomerUserError } from '@/types/shopify-storefront'

  export default mixins(signUp, signIn, clientStorageResetToken).extend ({
    name: 'SignUp',
    components: {
      FormProfile
    },
    metaInfo () {
      return {
        title: this.$t('signUp'),
        titleTemplate: '%s | Pawkan - Your best pet care'      
      }    
    },
    data: () => ({
      errorMessages: [] as CustomerUserError[]
    }),
    computed: {
      ...mapGetters([
        'prevNav'
      ])
    },
    methods: {
      ...mapMutations([
        'setCustomerAccessToken'
      ]),
      async submit (inputCustomerCreate: CustomerCreateInput): Promise<void> {
        const { email, password } = inputCustomerCreate
        const inputCustomerAccessTokenCreate =  { email, password }
        const result = await this.signUp_createUser(inputCustomerCreate, inputCustomerAccessTokenCreate)
        if (result.customerUserErrors.length) {
          this.errorMessages = result.customerUserErrors
        } else {
          const result = await this.signIn_createToken(inputCustomerAccessTokenCreate)                    
          if (result.customerUserErrors.length) {
            this.errorMessages = result.customerUserErrors
          } else
          if (result.customerAccessToken) {
            (this.$refs.formProfile as InstanceType<typeof FormProfile>).reset()
            this.setCustomerAccessToken(result.customerAccessToken.accessToken)
            this.clientStorageResetToken_reset(result.customerAccessToken.accessToken)    
            this.$router.push(this.prevNav)
          }
        }       
      }
    },
    beforeRouteEnter (...[, , next]) {
      if (store.getters.customerAccessToken) {
        next({ name: 'Home', params: { lang: 'en' } })
      } else {
        next()
      }
    }
  })
</script>