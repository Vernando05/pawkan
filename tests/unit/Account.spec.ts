import { renderComponent } from './config'
import Account from '@/views/Account.vue'
import SignUp from '@/views/SignUp.vue'
import SignIn from '@/views/SignIn.vue'

const routes = [
  { path: '/signin', name: 'SignIn', component: SignIn },
  { path: '/signup', name: 'SignUp', component: SignUp }
]

test('signout signin state', async () => {
  const store = {
    state: {},
    getters: {
      customerAccessToken: jest.fn(),
      currentLanguage: jest.fn()
    }
  }
  const { container, queryByText } = renderComponent(Account, {
    routes,
    store
  })
  expect(queryByText(/sign in/i)).toBeInTheDocument()
  expect(container).toMatchSnapshot()
})

test('signin state', async () => {
  const store = {
    state: {},
    getters: {
      customerAccessToken: jest.fn(() => 't0k3n'),
      currentLanguage: jest.fn()
    }
  }
  const { container, queryByText } = renderComponent(Account, {
    routes,
    store
  })
  expect(queryByText(/sign out/i)).toBeInTheDocument()
  expect(container).toMatchSnapshot()
})
