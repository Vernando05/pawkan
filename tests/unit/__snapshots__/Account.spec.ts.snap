// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signin state 1`] = `
<div
  data-app="true"
>
  <div>
    <!---->
     
    <div
      class="v-list v-sheet v-sheet--shaped theme--light"
      role="list"
    >
      <div
        class="v-item-group theme--light v-list-item-group primary--text"
        role="listbox"
      >
        <a
          aria-current="page"
          aria-selected="false"
          class="v-item--active v-list-item--active v-list-item v-list-item--link theme--light"
          href="#/"
          role="option"
          tabindex="0"
        >
          <div
            class="v-list-item__icon"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate mdi mdi-account theme--light"
            />
          </div>
           
          <div
            class="v-list-item__content"
          >
            <div
              class="v-list-item__title"
            >
              Profile
            </div>
          </div>
        </a>
         
        <div
          aria-selected="false"
          class="v-list-item v-list-item--link theme--light"
          role="option"
          tabindex="0"
        >
          <div
            class="v-list-item__icon"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate mdi mdi-logout-variant theme--light"
            />
          </div>
           
          <div
            class="v-list-item__content"
          >
            <div
              class="v-list-item__title"
            >
              Sign out
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`signout signin state 1`] = `
<div
  data-app="true"
>
  <div>
    <div
      class="row justify-center"
    >
      <div
        class="col"
      >
        <div
          class="mb-2"
        >
          <!---->
           
          <div
            class="mt-4 v-sheet theme--light"
          >
            <div
              aria-busy="true"
              aria-live="polite"
              class="v-skeleton-loader mx-auto mb-4 v-skeleton-loader--is-loading theme--light"
              role="alert"
              style="width: 220px;"
            >
              <div
                class="v-skeleton-loader__button v-skeleton-loader__bone"
              />
            </div>
             
            <div
              aria-busy="true"
              aria-live="polite"
              class="v-skeleton-loader mx-auto mb-4 v-skeleton-loader--is-loading theme--light"
              role="alert"
              style="width: 220px;"
            >
              <div
                class="v-skeleton-loader__button v-skeleton-loader__bone"
              />
            </div>
             
            <div
              aria-busy="true"
              aria-live="polite"
              class="v-skeleton-loader mx-auto mb-4 v-skeleton-loader--is-loading theme--light"
              role="alert"
              style="width: 220px;"
            >
              <div
                class="v-skeleton-loader__button v-skeleton-loader__bone"
              />
            </div>
             
            <div
              aria-busy="true"
              aria-live="polite"
              class="v-skeleton-loader mx-auto mb-4 v-skeleton-loader--is-loading theme--light"
              role="alert"
              style="width: 220px;"
            >
              <div
                class="v-skeleton-loader__button v-skeleton-loader__bone"
              />
            </div>
          </div>
           
          <div
            id="firebaseui-auth-container"
          />
           
          <div
            class="text-center"
            style="display: none;"
          >
            <a
              class="text-normal mb-4 elevation-3 justify-start v-btn v-btn--is-elevated v-btn--has-bg v-btn--router theme--light v-size--default main"
              data-testid="button-signin"
              href="#/signin"
              style="width: 220px;"
            >
              <span
                class="v-btn__content"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mr-3 mdi mdi-email theme--light"
                />
                Sign in with Email
              </span>
            </a>
             
            <a
              class="text-normal elevation-3 justify-start v-btn v-btn--has-bg v-btn--router theme--light v-size--default accent"
              href="#/signup"
              style="width: 220px;"
            >
              <span
                class="v-btn__content"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mr-3 mdi mdi-clipboard-account theme--light"
                />
                Sign up
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
     
    <div
      class="v-list v-sheet v-sheet--shaped theme--light"
      role="list"
    >
      <div
        class="v-item-group theme--light v-list-item-group primary--text"
        role="listbox"
      />
    </div>
  </div>
</div>
`;
